anesthetic/samples.py

```diff
@@ -482,6 +482,102 @@ class MCMCSamples(Samples):
+        nchains = chains.ngroups
+        if isinstance(burn_in, (int, float)) and 0 < burn_in < 1:
+            nsamples = chains.count().iloc[:, 0].to_numpy()
+            ndrop = np.ceil(burn_in * nsamples).astype(int)
+        elif isinstance(burn_in, (int, float)):
+            ndrop = np.full(nchains, burn_in, dtype=int)
+        elif isinstance(burn_in, (list, tuple, np.ndarray)) \
+                and len(burn_in) == nchains:
+            ndrop = burn_in
+        else:
+            raise ValueError("`burn_in` has to be a positive scalar or an "
+                             "array of length matching the number of chains "
+                             "`nchains=%d`. However, you provided "
+                             "`burn_in=%s`" % (nchains, burn_in))
+        data = self.drop(chains.apply(lambda g: g.head(ndrop[g.name-1])).index,
+                         inplace=inplace)
+        if reset_index:
+            data = data.reset_index(drop=True, inplace=inplace)
+        return data
+
+    def Gelman_Rubin(self, params=None):
+        """Gelman--Rubin convergence statistic of multiple MCMC chains.
+
+        Determine the Gelman--Rubin convergence statistic ``R-1`` by computing
+        and comparing the within-chain variance and the between-chain variance.
+        This follows the routine as outlined in Lewis (2013), section
+        IV.A., https://arxiv.org/abs/1304.4473.
+
+        Note that this requires more than one chain. To circumvent this, you
+        could overwrite the ``'chain'`` column, splitting the samples into two
+        or more sets.
+
+        Parameters
+        ----------
+        params: list(str)
+            List of column names (i.e. parameters) to be included in the
+            convergence calculation.
+            Default: all parameters (except those parameters that contain
+            'prior', 'chi2', or 'logL' in their names)
+
+        Returns
+        -------
+        Rminus1: float
+            Gelman--Rubin convergence statistic ``R-1``. The smaller, the
+            better converged. Aiming for ``Rminus1~0.01`` should normally work
+            well.
+        """
+        self.columns.set_names(['params', 'labels'], inplace=True)
+        if params is None:
+            params = [key for key in self.columns.get_level_values('params')
+                      if 'prior' not in key
+                      and 'chi2' not in key
+                      and 'logL' not in key
+                      and 'chain' not in key]
+        chains = self[params+['chain']].groupby(
+                ('chain', '$n_\\mathrm{chain}$')
+        )
+        nchains = chains.ngroups
+
+        # Within chain variance ``W``
+        # (average variance within each chain):
+        covs = chains.cov()
+        W = covs.groupby(level=['params', 'labels']).mean().to_numpy()
+        # TODO: the above line should be a weighted mean
+        # --> need to fix groupby for WeightedDataFrames!
```

Comment:
Thanks to our `WeightedDataFrames` the computation of the Gelman--Rubin statistic is really quite neat. There is this one issue that the data frames resulting from a `groupby` action should in general be weighted data frames themselves, which they currently are not.

A simple example is the mean of a parameter across the full weighted dataset should be the same as the (weighted!) mean of the means of all individual chains, i.e. we should have (in slightly simplified syntax for readability):
```python
assert samples.groupby('chain').mean().mean() == samples.mean()
```
Note the double `.mean()`, the first produces the within chain means that themselves should be weighted, and the mean of these chain means should match the overall mean.

Reply:
We could think about an optional `skipna` kwarg, but yes, by default I think it should be `skipna=True` which should result in dropping rows. For that particular case I'd be happy to accept a discrepancy between the full mean and the groupby mean.

Reply:
Should I turn this into an issue for now, to be addressed in a PR targeted at `groupby`? 

While the current situation isn't perfect, I think in practice for sufficiently long chains this shouldn't matter quite that much.

Reply:
>Should I turn this into an issue for now, to be addressed in a PR targeted at groupby?

Please do

Reply:
Done in #260.